skops/io/_numpy.py

```diff
@@ -15,47 +16,53 @@ def ndarray_get_state(obj, dst):
         "__module__": get_module(type(obj)),
     }
 
-    try:
+    # First, try to save object with np.save and allow_pickle=False, which
+    # should generally work as long as the dtype is not object.
+    with suppress(ValueError):
         f_name = f"{uuid4()}.npy"
         with open(Path(dst) / f_name, "wb") as f:
             np.save(f, obj, allow_pickle=False)
-            res["type"] = "numpy"
-            res["file"] = f_name
-    except ValueError:
-        # Object arrays cannot be saved with allow_pickle=False, therefore we
-        # convert them to a list and recursively call get_state on it.
-        if obj.dtype == object:
-            obj_serialized = get_state(obj.tolist(), dst)
-            res["content"] = obj_serialized["content"]
-            res["type"] = "json"
-            res["shape"] = get_state(obj.shape, dst)
-        else:
-            raise TypeError(f"numpy arrays of dtype {obj.dtype} are not supported yet")
-
+        res.update(type="numpy", file=f_name)
+        return res
+
+    # Object arrays cannot be saved with allow_pickle=False, therefore we
+    # convert them to a list and recursively call get_state on it. For this, we
+    # expect the dtype to be object.
+    if obj.dtype != object:
+        raise UnsupportedTypeException(
```

Comment:
I tried to write a test for it but it wasn't that simple. This is the easiest I could come up with:

```python
class _MockDtype:
    metadata = None
    names = None
    subdtype = None
    str = "<f8"
    hasobject = False

_mock_dtype = _MockDtype()


class _CustomDtypeArray(np.ndarray):
    @property
    def dtype(self):
        return _mock_dtype

    @property
    def itemsize(self):
        # np.save checks this attribute, so (ab)use this to trigger a ValueError
        raise ValueError("Trigger ValueError")


class NumpyUnknownDtypeEstimator(BaseEstimator):
    def fit(self, X, y=None, **fit_params):
        self.x_ = _CustomDtypeArray([1, 2, 3])
        return self


def test_unknown_numpy_dtype_raises(tmp_path):
    est = NumpyUnknownDtypeEstimator().fit(None)
    f_name = tmp_path / "file.skops"
    with pytest.raises(UnsupportedTypeException):
        save_load_round(est, f_name)
```

I found it interesting that without the `itemsize` property, saving would actually just work! Loading would fail though.

But I think what this shows us is that it's almost impossible to trigger this code path as a user, even with custom dtypes (if that's even a thing), unless the user goes throw extraordinary lengths to provoke the error. So I don't think this situation will really come up in the wild.

This makes me think that we might have to check the dtype differently inside this function. Not sure if we want to support custom dtypes and hope for the best or if we should check against all existing dtypes and raise if we don't know this one.

Reply:
sklearn has custom dtypes, in trees, but if that's not triggering this, then not many things probably would, and that means we can remove it?

Reply:
Thanks for the pointer. I checked and `np.save` and `np.load` seem to work fine on that custom dtype (I suppose you mean this `dtype=[('left_child', '<i8'), ...]` thing).

It means we could remove the `if obj.dtype != object:` check, given how we never trigger it. I just wonder if it isn't more cautious to leave it there to have an explicit error? We don't really know what would happen if we have a custom dtype that cannot be saved with `np.save`, maybe it's better to raise and get a user report? We could encourage reporting by extending the error message with "Please open an issue on...".

Reply:
Yeah, extending the error message makes sense. I'm okay with this line not being tested.